File: tutorials/tour-of-restate-rust/src/part1/ticket_object.rs
use crate::auxiliary::ticket_status::TicketStatus;
use restate_sdk::prelude::*;

#[restate_sdk::object]
pub(crate) trait TicketObject {
    async fn reserve() -> Result<bool, HandlerError>;
    async fn unreserve() -> Result<(), HandlerError>;
    #[name = "markAsSold"]
    async fn mark_as_sold() -> Result<(), HandlerError>;
}

pub struct TicketObjectImpl;

const STATUS: &str = "status";

impl TicketObject for TicketObjectImpl {
    async fn reserve(&self, ctx: ObjectContext<'_>) -> Result<bool, HandlerError> {
        Ok(true)
    }

    async fn unreserve(&self, ctx: ObjectContext<'_>) -> Result<(), HandlerError> {
        Ok(())
    }

    async fn mark_as_sold(&self, ctx: ObjectContext<'_>) -> Result<(), HandlerError> {
        Ok(())
    }
}

File: go/tutorials/tour-of-restate-go/part4/ticketobject.go
package main

import (
	"github.com/restatedev/examples/tutorials/tour-of-restate-go/auxiliary"
	restate "github.com/restatedev/sdk-go"
)

type TicketObject struct{}

func (TicketObject) Reserve(ctx restate.ObjectContext) (bool, error) {
	status, err := restate.Get[auxiliary.TicketStatus](ctx, "status")
	if err != nil {
		return false, err
	}

	if status == auxiliary.TicketStatusAvailable {
		restate.Set(ctx, "status", auxiliary.TicketStatusReserved)
		return true, nil
	} else {
		return false, nil
	}
}

func (TicketObject) Unreserve(ctx restate.ObjectContext) error {
	status, err := restate.Get[auxiliary.TicketStatus](ctx, "status")
	if err != nil {
		return err
	}

	if status != auxiliary.TicketStatusSold {
		restate.Clear(ctx, "status")
	}

	return nil
}

func (TicketObject) MarkAsSold(ctx restate.ObjectContext) error {
	status, err := restate.Get[auxiliary.TicketStatus](ctx, "status")
	if err != nil {
		return err
	}

	if status == auxiliary.TicketStatusReserved {
		restate.Set(ctx, "status", auxiliary.TicketStatusSold)
	}

	return nil
}
